File: ueb6/app/Messages.hs
{-# LANGUAGE OverloadedStrings #-}

module Messages where

import qualified Data.Map.Lazy as M
import Data.Aeson
import OT

data ClientMessage = ClientEdit Int TextOperation Cursor

data ServerMessage = RemoteEdit TextOperation (M.Map Int Cursor)
                   | Ack
                   | HelloGuys

instance FromJSON TextAction where
  parseJSON (Number x) = do
    n <- parseJSON (Number x)
    case (n :: Int) of
      -1  -> return Delete
      1   -> return Retain
      o   -> error $ "num: " ++ show o
  parseJSON (String t) = do
    c <- parseJSON (String t)
    return $ Insert c

instance FromJSON TextOperation where
  parseJSON (Array a) = do
    ops <- parseJSON (Array a)
    return $ TextOperation ops

instance FromJSON ClientMessage where
  parseJSON (Object o) = ClientEdit
    <$> o .: "rev"
    <*> o .: "op"
    <*> o .: "cursor" 

instance ToJSON TextAction where
  toJSON Retain = toJSON (1 :: Int)
  toJSON Delete = toJSON (-1 :: Int)
  toJSON (Insert c) = toJSON c

instance ToJSON TextOperation where
  toJSON (TextOperation as) = toJSON $ map toJSON as

instance ToJSON ServerMessage where
  toJSON Ack = toJSON ("ack" :: String)
  toJSON HelloGuys = toJSON ("helloguys" :: String)
  toJSON (RemoteEdit op cursors) = object ["op" .= toJSON op, 
    "cursors" .= cursors'] where cursors' = toJSON $ M.elems cursors 
 

File: ueb6/client/src/main/scala/Messages.scala
import io.circe.{Decoder, DecodingFailure, Encoder, Json}
import io.circe.syntax._
import org.scalajs.dom.{document, html, console}

sealed trait ClientMessage
case class ClientEdit(
  rev: Int,
  op: TextOperation.Operation,
  cursor: Int) extends ClientMessage

sealed trait ServerMessage
case object Ack extends ServerMessage
case object HelloGuys extends ServerMessage 
case class RemoteEdit(
  op: TextOperation.Operation,
  cursors: List[Int]
) extends ServerMessage

object ClientMessage {
  implicit val encodeAction =
    Encoder.instance[TextAction] {
      case Retain(n) => n.asJson
      case Insert(cs) => cs.asJson
      case Delete(n) => (-n).asJson
    }

  implicit val encodeClientMessage = Encoder.instance[ClientMessage] {
    case ClientEdit(rev,op,cursor) => Json.obj(
      "rev" -> rev.asJson,
      "op" -> op.asJson,
      "cursor" -> cursor.asJson
    )
  }
}

object ServerMessage {
  implicit val parseAction =
    Decoder.decodeString.map[TextAction](Insert) or
    Decoder.decodeInt.map[TextAction] { n => 
      if (n < 0)
        Delete(-n)
      else  
        Retain(n)
    }

  val parseAckAndHello = Decoder.decodeString.emap[ServerMessage] {
    case "ack" => Right(Ack)
    case "helloguys" => Right(HelloGuys)
    case other => Left("invalid server message")
  }


  val parseRemoteEdit : Decoder[ServerMessage] = 
    Decoder.instance( o =>
      for {
        op <- o.downField("op").as[List[TextAction]] 
        cursors <- o.downField("cursors").as[List[Int]]
      } yield RemoteEdit(op,cursors)
    )

  implicit val parseServerMessage =
    parseAckAndHello or parseRemoteEdit
}
